iceoryx2-pal/concurrency-sync/src/ice_atomic.rs

```diff
@@ -0,0 +1,356 @@
+pub mod internal {
+    use core::ops::BitAnd;
+
+    use super::*;
+
+    pub trait AtomicInteger:
+        Copy
+        + Send
+        + Eq
+        + AddAssign
+        + SubAssign
+        + BitAndAssign
+        + BitOrAssign
+        + BitXorAssign
+        + BitAnd<Output = Self>
+        + Ord
+        + Not<Output = Self>
+    {
+        fn overflowing_add(self, rhs: Self) -> (Self, bool);
+        fn overflowing_sub(self, rhs: Self) -> (Self, bool);
+    }
+
+    impl AtomicInteger for u64 {
+        fn overflowing_add(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_add(rhs)
+        }
+
+        fn overflowing_sub(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_sub(rhs)
+        }
+    }
+
+    impl AtomicInteger for u128 {
+        fn overflowing_add(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_add(rhs)
+        }
+
+        fn overflowing_sub(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_sub(rhs)
+        }
+    }
+
+    impl AtomicInteger for i64 {
+        fn overflowing_add(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_add(rhs)
+        }
+
+        fn overflowing_sub(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_sub(rhs)
+        }
+    }
+
+    impl AtomicInteger for i128 {
+        fn overflowing_add(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_add(rhs)
+        }
+
+        fn overflowing_sub(self, rhs: Self) -> (Self, bool) {
+            self.overflowing_sub(rhs)
+        }
+    }
+}
+
+/// iceoryx2 implementation of an atomic that has an internal [`RwLockWriterPreference`].
+/// It enables atomic operations on platforms that do not support them with the restriction that
+/// those operations are no longer lock-free.
+#[repr(C)]
+pub struct IceAtomic<T: internal::AtomicInteger> {
+    data: UnsafeCell<T>,
+    lock: LockType,
+}
+
+impl<T: internal::AtomicInteger> IceAtomic<T> {
+    /// See [`core::sync::atomic::AtomicU64::new()`]
+    pub fn new(v: T) -> Self {
+        Self {
+            data: UnsafeCell::new(v),
+            lock: LockType::new(),
+        }
+    }
+
+    fn read_lock(&self) {
+        self.lock.read_lock(|_, _| WaitAction::Continue);
+    }
+
+    fn write_lock(&self) {
+        self.lock
+            .write_lock(|_, _| WaitAction::Continue, |_| {}, |_| {});
+    }
+
+    fn unlock(&self) {
+        self.lock.unlock(|_| {}, |_| {});
+    }
+
+    /// See [`core::sync::atomic::AtomicU64::as_ptr()`]
+    pub const fn as_ptr(&self) -> *mut T {
+        self.data.get()
+    }
+
+    /// See [`core::sync::atomic::AtomicU64::compare_exchange()`]
+    pub fn compare_exchange(
+        &self,
+        current: T,
+        new: T,
+        _success: Ordering,
+        _failure: Ordering,
+    ) -> Result<T, T> {
+        self.write_lock();
+        let data = unsafe { *self.data.get() };
+        if data != current {
+            core::sync::atomic::fence(Ordering::SeqCst);
```

Comment:
Shouldn't the lock take care of the synchronization?

Reply:
The lock is based on an `AtomicU32` with acquire release semantics. When the user wants to have sequential consistency it is not enough. And for 32-bit I would make everything sequentially consistent in the "compatibility" ice atomic

Reply:
I'm not sure that throwing fences everywhere is a good idea. If the lock does not properly synchronize then it is broken and needs to be fixed, else one would need to add those fences to all the places it is used. It might even be a good chance to test the look with these atomics. Unfortunately it won't be easy to get a native 32 bit target to the CI

Reply:
The reason for the `seq_cst` fence is that a user's algorithm may depend on being sequentially consistent, but the RwLock has only acquired release memory order.

And instead of thinking everything through to the last detail how what where can how interact, I thought: be safe use a sequentially consistent memory barrier.

Do you think this can cause bugs or problems when we also assume that 32-bit support is only tier 2?

Reply:
That depends how we define tier 2 but at some point it needs to become lock-free